eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWits.hs

```diff
@@ -579,55 +610,75 @@ instance
-                )
+                nonEmptySetDecoder
                 (mapTraverseableDecoderA (decodeList decCBOR) Set.fromList)
           )
       txWitnessField 1 =
-        fieldAA
-          addScriptsTxWitsRaw
-          (D nativeScriptsDecoder)
+        fieldAA addScriptsTxWitsRaw (D nativeScriptsDecoder)
       txWitnessField 2 =
         fieldAA
           (\x wits -> wits {atwrBootAddrTxWits = x})
           ( D $
               ifDecoderVersionAtLeast
                 (natVersion @9)
-                ( allowTag setTag
-                    >> mapTraverseableDecoderA (decodeNonEmptyList decCBOR) (Set.fromList . NE.toList)
-                )
+                nonEmptySetDecoder
                 (mapTraverseableDecoderA (decodeList decCBOR) Set.fromList)
           )
       txWitnessField 3 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV1)
-      txWitnessField 4 =
-        fieldAA
-          (\x wits -> wits {atwrDatsTxWits = x})
-          From
+      txWitnessField 4 = fieldAA (\x wits -> wits {atwrDatsTxWits = x}) From
       txWitnessField 5 = fieldAA (\x wits -> wits {atwrRdmrsTxWits = x}) From
       txWitnessField 6 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV2)
       txWitnessField 7 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV3)
       txWitnessField n = invalidField n
       {-# INLINE txWitnessField #-}
 
+      hashedNativeSciptDecoder :: Decoder s (Annotator (ScriptHash, Script era))
+      hashedNativeSciptDecoder = fmap (asHashedScriptPair @era . fromNativeScript) <$> decCBOR
+      {-# INLINE hashedNativeSciptDecoder #-}
+
+      noDuplicateNativeScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
+      noDuplicateNativeScriptsDecoder =
+        noDuplicateNonEmptySetAsMapDecoderAnn
+          (\ns -> (hashScript $ fromNativeScript ns, fromNativeScript ns))
+      {-# INLINE noDuplicateNativeScriptsDecoder #-}
+
       nativeScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
       nativeScriptsDecoder =
         ifDecoderVersionAtLeast
-          (natVersion @9)
-          ( allowTag setTag
-              >> mapTraverseableDecoderA (decodeNonEmptyList pairDecoder) (Map.fromList . NE.toList)
+          (natVersion @12)
+          noDuplicateNativeScriptsDecoder
+          ( do
+              allowTag setTag
+              res <- mapTraverseableDecoderA (decodeList hashedNativeSciptDecoder) Map.fromList
+              pure $ ensureNonEmpty res
           )
-          (mapTraverseableDecoderA (decodeList pairDecoder) Map.fromList)
         where
-          pairDecoder :: Decoder s (Annotator (ScriptHash, Script era))
-          pairDecoder = fmap (asHashedScriptPair . fromNativeScript) <$> decCBOR
+          ensureNonEmpty :: Annotator (Map ScriptHash (Script era)) -> Annotator (Map ScriptHash (Script era))
+          ensureNonEmpty x = do
+            x' <- x
+            when (null x') $ fail "Map not empty"
+            x
```

Comment:
Couple of things about this function:
* On the wire it is a Set
* we fail when it is empty, not when it is "not empty" :grin:
* Avoiding ticks is a good practice.
* you already run `x` action once, which means it is better to return the result, instead of running it again.
```suggestion
          ensureNonEmpty xAnn = do
            x <- xAnn
            when (null x) $ fail "Set is empty"
            pure x
```